Sources/Networking/Core/DownloadAPIManager.swift

```diff
@@ -0,0 +1,193 @@
+            downloadTask.resume()
+            
+            updateTasks()
+            
+            let urlResponse = try await downloadTask.asyncResponse()
+            
+            /// process response
+            let response = try await responseProcessors.process((Data(), urlResponse), with: request, for: endpointRequest)
+            
+            /// reset retry count
+            await retryCounter.reset(for: endpointRequest.id)
+            
+            /// create download AsyncStream
+            return (downloadTask, response)
+        } catch {
+            do {
+                /// If retry fails (retryCount is 0 or Task.sleep thrown), catch the error and process it with `ErrorProcessing` plugins.
+                try await sleepIfRetry(
+                    for: error,
+                    endpointRequest: endpointRequest,
+                    retryConfiguration: retryConfiguration
+                )
+                
+                return try await downloadRequest(
+                    endpointRequest,
+                    resumableData: resumableData,
+                    retryConfiguration: retryConfiguration
+                )
+            } catch {
+                /// error processing
+                throw await errorProcessors.process(error, for: endpointRequest)
+            }
+        }
+    }
+    
+    /// Creates a record in the `downloadStateDict` for each task and observes their states.
+    /// Every `downloadStateDict` update triggers an event to the `downloadStateDictSubject`
+    /// which then leads to a task state update from `progressStream`.
+    func updateTasks() {
+        Task {
+            for task in await allTasks where downloadStateDict[task] == nil {
+                /// In case there is no DownloadState for a given task in the dictionary, we need to create one.
+                downloadStateDict[task] = .init(task: task)
+                
+                /// We need to observe URLSessionTask.State via KVO individually for each task, because there is no delegate callback for the state change.
+                taskStateCancellables[task] = task
+                    .publisher(for: \.state)
+                    .sink { [weak self] state in
+                        self?.downloadStateDict[task]?.taskState = state
+                        
+                        if state == .completed {
+                            self?.taskStateCancellables[task] = nil
+                        }
+                    }
+            }
+        }
+    }
+}
+
+// MARK: URLSession Delegate
+extension DownloadAPIManager: URLSessionDelegate, URLSessionDownloadDelegate {
+    public func urlSession(_: URLSession, downloadTask: URLSessionDownloadTask, didWriteData _: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
+        downloadStateDict[downloadTask]?.downloadedBytes = totalBytesWritten
+        downloadStateDict[downloadTask]?.totalBytes = totalBytesExpectedToWrite
+    }
```

Comment:
Could this be the place  where (accessing the `downloadStateDict`) it's not thread-safe and perhaps causing the crash? Maybe using an actor would be safer.

Reply:
👍 similar we have for running calls and retry mechanism

Reply:
agreed, all props which are being mutated should be thread safe